src/project/solve_group.rs

```diff
@@ -65,122 +57,35 @@ impl<'p> SolveGroup<'p> {
-                // Either clone the values from the Cow or move the values from the owned map.
-                let deps_iter = match deps {
-                    Cow::Borrowed(borrowed) => Either::Left(
-                        borrowed
-                            .into_iter()
-                            .map(|(name, spec)| (name.clone(), spec.clone())),
-                    ),
-                    Cow::Owned(owned) => Either::Right(owned.into_iter()),
-                };
-
-                // Add the requirements to the accumulator.
-                for (name, spec) in deps_iter {
-                    acc.entry(name).or_default().push(spec);
-                }
-
-                acc
-            })
+        self.local_system_requirements()
     }
+}
 
-    /// Returns the channels associated with this solve group.
-    ///
-    /// Users can specify custom channels on a per-feature basis. This method collects and
-    /// deduplicates all the channels from all the features in the order they are defined in the
-    /// manifest.
+impl<'p> CombineFeature<'p> for SolveGroup<'p> {
+    /// Returns all features that are part of the solve group.
     ///
-    /// If a feature does not specify any channel the default channels from the project metadata are
-    /// used instead. However, these are not considered during deduplication. This means the default
-    /// channels are always added to the end of the list.
-    pub fn channels(&self) -> IndexSet<&'p Channel> {
-        self.features(true)
-            .filter_map(|feature| match feature.name {
-                // Use the user-specified channels of each feature if the feature defines them. Only
-                // for the default feature do we use the default channels from the project metadata
-                // if the feature itself does not specify any channels. This guarantees that the
-                // channels from the default feature are always added to the end of the list.
-                FeatureName::Named(_) => feature.channels.as_deref(),
-                FeatureName::Default => feature
-                    .channels
-                    .as_deref()
-                    .or(Some(&self.project.manifest.parsed.project.channels)),
-            })
-            .flatten()
-            // The prioritized channels contain a priority, sort on this priority.
-            // Higher priority comes first. [-10, 1, 0 ,2] -> [2, 1, 0, -10]
-            .sorted_by(|a, b| {
-                let a = a.priority.unwrap_or(0);
-                let b = b.priority.unwrap_or(0);
-                b.cmp(&a)
-            })
-            .map(|prioritized_channel| &prioritized_channel.channel)
-            .collect()
+    /// All features of all environments are combined and deduplicated.
+    fn features(&self) -> impl DoubleEndedIterator<Item = &'p manifest::Feature> + 'p {
+        self.environments()
+            .flat_map(|env: Environment<'p>| env.features().collect_vec().into_iter())
```

Comment:
I had to add the `.collect_vec().into_iter()` here.  The logic seems exactly the same as before, but for the encapsulation in the trait. For the life of me I could not make it work otherwise 😱 . If you rust gurus can give me a hint....

Reply:
Yeah, couldn't get this to work either. Which seems weird to me because all lifetimes should be the same. But let's leave it.

Reply:
OK cool. I feel better knowing this :-D